File: frontend/src/types/report.ts
/**
 * Report types for sustainability report analysis.
 */

import type { Claim, ClaimVerdict } from "./claim";
import type { AgentFinding } from "./agent";
import type { DisclosureGap, PillarSummary } from "./ifrs";

export type ReportStatus =
  | "uploaded"
  | "parsing"
  | "embedding"
  | "parsed"
  | "analyzing"
  | "completed"
  | "error";

export interface Report {
  id: string;
  filename: string;
  fileSizeBytes: number;
  pageCount: number | null;
  status: ReportStatus;
  errorMessage: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface SectionInfo {
  title: string;
  level: number;
  page_start: number;
  page_end: number | null;
  children: SectionInfo[];
}

export interface ContentStructure {
  sections: SectionInfo[];
  table_count: number;
  page_count: number;
  estimated_word_count: number;
}

export interface UploadResponse {
  report_id: string;
  filename: string;
  file_size_bytes: number;
  status: ReportStatus;
  created_at: string;
}

export interface ReportStatusResponse {
  report_id: string;
  filename: string;
  file_size_bytes: number;
  status: ReportStatus;
  page_count: number | null;
  content_structure: ContentStructure | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface SourceOfTruthReport {
  report: Report;
  claims: ClaimWithVerdict[];
  pillarSummaries: PillarSummary[];
  disclosureGaps: DisclosureGap[];
  generatedAt: string;
}

export interface ClaimWithVerdict {
  claim: Claim;
  verdict: ClaimVerdict | null;
  findings: AgentFinding[];
}

File: backend/app/models/report.py
"""Report model - represents an uploaded sustainability report."""

from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import BigInteger, DateTime, Index, LargeBinary, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base, generate_uuid7

if TYPE_CHECKING:
    from app.models.chat import Conversation
    from app.models.claim import Claim
    from app.models.embedding import Embedding
    from app.models.finding import Finding
    from app.models.verdict import Verdict


class Report(Base):
    """
    Represents an uploaded sustainability report.

    Status values:
    - uploaded: PDF received, not yet parsed
    - parsing: PyMuPDF4LLM extraction in progress
    - embedding: Content parsed, embedding chunks via RAG service
    - parsed: Content extracted and embedded, ready for analysis
    - analyzing: Agent pipeline running
    - completed: Full pipeline complete, Source of Truth available
    - error: Processing failed (see error_message)
    """

    __tablename__ = "reports"

    id: Mapped[UUID] = mapped_column(
        primary_key=True,
        default=generate_uuid7,
    )
    filename: Mapped[str] = mapped_column(String(255), nullable=False)
    file_size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    page_count: Mapped[int | None] = mapped_column(nullable=True)
    status: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default="uploaded",
    )
    parsed_content: Mapped[str | None] = mapped_column(Text, nullable=True)
    pdf_binary: Mapped[bytes | None] = mapped_column(LargeBinary, nullable=True)
    content_structure: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    # Relationships
    claims: Mapped[list["Claim"]] = relationship(
        "Claim",
        back_populates="report",
        cascade="all, delete-orphan",
    )
    embeddings: Mapped[list["Embedding"]] = relationship(
        "Embedding",
        back_populates="report",
        cascade="all, delete-orphan",
    )
    findings: Mapped[list["Finding"]] = relationship(
        "Finding",
        back_populates="report",
        cascade="all, delete-orphan",
    )
    verdicts: Mapped[list["Verdict"]] = relationship(
        "Verdict",
        back_populates="report",
        cascade="all, delete-orphan",
    )
    conversation: Mapped["Conversation | None"] = relationship(
        "Conversation",
        back_populates="report",
        cascade="all, delete-orphan",
        uselist=False,
    )

    __table_args__ = (
        Index("ix_reports_status", "status"),
    )
